presto-main/src/main/java/com/facebook/presto/operator/LookupJoinPageBuilder.java

```diff
@@ -0,0 +1,140 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.facebook.presto.operator;
+
+import com.facebook.presto.spi.Page;
+import com.facebook.presto.spi.PageBuilder;
+import com.facebook.presto.spi.block.Block;
+import com.facebook.presto.spi.type.Type;
+import it.unimi.dsi.fastutil.ints.IntArrayList;
+
+import java.util.List;
+
+import static com.facebook.presto.spi.block.PageBuilderStatus.DEFAULT_MAX_PAGE_SIZE_IN_BYTES;
+import static com.google.common.base.MoreObjects.toStringHelper;
+import static com.google.common.base.Verify.verify;
+
+/**
+ * This page builder creates pages with dictionary blocks:
+ * normal dictionary blocks for the probe side and the original blocks for the build side.
+ *
+ * TODO use dictionary blocks (probably extended kind) to avoid data copying for build side
+ */
+public class LookupJoinPageBuilder
+{
+    private final IntArrayList probeIndexBuilder = new IntArrayList();
+    private final PageBuilder buildPageBuilder;
+    private final int buildOutputChannelCount;
+    private int probeBlockBytes;
+
+    public LookupJoinPageBuilder(List<Type> buildTypes)
+    {
+        this.buildPageBuilder = new PageBuilder(buildTypes);
```

Comment:
You could `requireNonNull(buildTypes)` here (i know LJO does that, I'd have it in both places)